File: 1129-longest-string-chain/longest-string-chain.java
class Solution {
    public boolean solve(String w1,String w2){
        int i = 0, j = 0;
        while (i < w1.length() && j < w2.length()) {
            if (w1.charAt(i) == w2.charAt(j)) {
                i++;
            }
            j++;
        }
        return i == w1.length();
    }
    public int longestStrChain(String[] words) {
        if (words == null || words.length == 0) return 0;
        int n=words.length;
        int[] dp=new int[n];
        Arrays.fill(dp,1);
        int maxi=1;
        Arrays.sort(words,Comparator.comparingInt(String::length));
        for(int i=1;i<n;i++){
            for(int j=0;j<i;j++){
                if (words[i].length() - words[j].length() == 1 && solve(words[j], words[i]) && dp[j]+1>dp[i])dp[i]=dp[j]+1;
            }
            if(dp[i]>maxi)maxi=dp[i];
        }
        return maxi;
    }
}

File: 1129-longest-string-chain/longest-string-chain.py
class Solution(object):
    def longestStrChain(self, word):
        """
        :type words: List[str]
        :rtype: int
        """
        if not word:
            return 0
        word.sort(key=len)
        def solve(w1,w2):
            if len(w2) - len(w1) != 1:
                return False
            i = j = 0
            while i < len(w1) and j < len(w2):
                if w1[i] == w2[j]:
                    i += 1
                j += 1
            return i == len(w1)
        dp = [1] * len(word)
        maxi = 1
        for i in range(1,len(word)):
            for j in range(0,i):
                if(len(word[i]) - len(word[j]) == 1
                    and solve(word[j], word[i])and dp[j]+1>dp[i]):
                    dp[i]=dp[j]+1
            if(dp[i]>maxi):
                maxi=dp[i]
        return maxi      